File: Python/2_Add Two Numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        l3 = ListNode(0, None)
        aux1 = l1
        aux2 = l2
        aux3 = l3
        vai1 = False
        while (aux1 or aux2):
            if aux1:
                aux3.val += aux1.val
            if aux2:
                aux3.val += aux2.val
            if vai1 == True: #Verifica se tem mais 1 da ultima iteracao
                aux3.val += 1
                vai1 = False
            if aux3.val > 9: #Verifica se tem mais 1 na proxima iteracao
                aux3.val = aux3.val % 10
                vai1 = True
            print(aux1, aux2, aux3, vai1, sep="\n")
            if aux1: #Avança aux1 se não for None
                aux1 = aux1.next
            if aux2: #Avança aux2 se não for None
                aux2 = aux2.next
            if aux1 or aux2:
                aux3.next = ListNode(0,None)
            if not aux1 and not aux2 and vai1 == True:
                aux3.next = ListNode(1,None)
            aux3 = aux3.next
            print("-------------------")
        return l3

            

File: Go/2_Add Two Numbers.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	result := new(ListNode)
	result.Val = 0
	result.Next = nil

	auxL1 := l1
	auxL2 := l2
	auxResult := result

	for {
		if auxL1 != nil {
			auxResult.Val += auxL1.Val
			auxL1 = auxL1.Next
		}
		if auxL2 != nil {
			auxResult.Val += auxL2.Val
			auxL2 = auxL2.Next
		}
		if auxResult.Val >= 10 {
			auxResult.Next = new(ListNode)
			auxResult.Val = auxResult.Val - 10
			auxResult.Next.Val = 1
		}
		if (auxL1 != nil || auxL2 != nil) && auxResult.Next == nil {
			auxResult.Next = new(ListNode)
			auxResult.Next.Val = 0
		}
		auxResult = auxResult.Next
		if auxL1 == nil && auxL2 == nil {
			break
		}
	}

	return result
}
